Blocks/HTTP/impl/posix_server.h

```diff
@@ -298,32 +362,49 @@ class HTTPServerPOSIX final {
         // Silently discard errors if no data was sent in.
       } catch (const current::Exception& e) {  // LCOV_EXCL_LINE
         // TODO(dkorolev): More reliable logging.
-        std::cerr << "HTTP route failed: " << e.what() << "\n";  // LCOV_EXCL_LINE
+        std::cerr << "HTTP route failed: " << e.what() << '\n';  // LCOV_EXCL_LINE
       }
     }
   }
 
   HTTPRoutesScopeEntry DoRegisterHandler(const std::string& path,
                                          std::function<void(Request)> handler,
-                                         const URLPathParams::Count path_param_count,
+                                         const URLPathArgs::CountMask path_param_count_mask,
                                          const ReRegisterRoute policy) {
-    if (path.empty() || path[0] != '/' || (path.length() > 1 && path[path.length() - 1] == '/') ||
-        !URL::IsValidPath(path)) {
-      CURRENT_THROW(InvalidHandlerPathException("Invalid handler path '" + path + "'"));
+    assert(handler);
+
+    if (path.empty() || path[0] != '/') {
+      CURRENT_THROW(PathDoesNotStartWithSlash("HTTP path does not start with a slash: `" + path + "`."));
+    }
+    if (path.length() > 1 && path[path.length() - 1] == '/') {
+      CURRENT_THROW(PathEndsWithSlash("HTTP path ends with slash: `" + path + "`."));
     }
+    if (!URL::IsPathValidToRegister(path)) {
+      CURRENT_THROW(PathContainsInvalidCharacters("HTTP path contains invalid characters: `" + path + "`."));
+    }
+
+    auto& handlers_per_path = handlers_[path];
+    URLPathArgs::CountMask mask = URLPathArgs::CountMask::None;
+    for (size_t i = 0; i <= MaximumUrlPathParams(); ++i, mask = mask << 1) {
+      if ((path_param_count_mask & mask) == mask) {
+        // Check if handler with exactly the same path exists and overwrite it if policy allows.
+        auto& placeholder = handlers_per_path[i];
```

Comment:
If you're talking about reverting just in case (i.e., leave the state unchanged if the user caught the exception), I was just falling asleep yesterday. `git push -f` coming coon. :-)